sdk/src/sdk/vdb/grpc/kuksa_val_v2/BrokerClient.cpp

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (c) 2022-2024 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "BrokerClient.h"
+
+#include "sdk/DataPointValue.h"
+#include "sdk/Logger.h"
+#include "sdk/middleware/Middleware.h"
+#include "sdk/vdb/grpc/kuksa_val_v2/BrokerAsyncGrpcFacade.h"
+#include "sdk/vdb/grpc/kuksa_val_v2/TypeConversions.h"
+
+#include <fmt/core.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+
+#include <limits>
+#include <stdexcept>
+#include <utility>
+
+namespace velocitas::kuksa_val_v2 {
+
+namespace {
+
+int assertProtobufArrayLimits(size_t numElements) {
+    if (numElements > std::numeric_limits<int>::max()) {
+        throw std::runtime_error("# requested datapoints exceeds gRPC limits");
+    }
+    return static_cast<int>(numElements);
+}
+
+} // namespace
+
+BrokerClient::BrokerClient(const std::string& vdbAddress, std::string vdbServiceName) {
+    logger().info("Connecting to data broker service '{}' via '{}'", vdbServiceName, vdbAddress);
+    m_asyncBrokerFacade = std::make_shared<BrokerAsyncGrpcFacade>(
+        grpc::CreateChannel(vdbAddress, grpc::InsecureChannelCredentials()));
+    Middleware::Metadata metadata = Middleware::getInstance().getMetadata(vdbServiceName);
+    m_asyncBrokerFacade->setContextModifier([metadata](auto& context) {
+        for (auto metadatum : metadata) {
+            context.AddMetadata(metadatum.first, metadatum.second);
+        }
+    });
+}
+
+BrokerClient::BrokerClient(const std::string& vdbServiceName)
+    : BrokerClient(Middleware::getInstance().getServiceLocation(vdbServiceName), vdbServiceName) {}
+
+BrokerClient::~BrokerClient() {}
+
+AsyncResultPtr_t<DataPointReply>
+BrokerClient::getDatapoints(const std::vector<std::string>& paths) {
+    auto result = std::make_shared<AsyncResult<DataPointReply>>();
+
+    kuksa::val::v2::GetValuesRequest request;
+
+    auto& signalIds = *request.mutable_signal_ids();
+
+    signalIds.Reserve(assertProtobufArrayLimits(paths.size()));
+    for (const auto& path : paths) {
+        signalIds.Add()->set_path(path);
+    }
+
+    m_asyncBrokerFacade->GetValues(
+        std::move(request),
+        [result, paths](auto reply) {
```

Comment:
`const auto& reply` ?

Reply:
Not sure - because the function receiving the callback expects a function expecting a value - no reference